wip for legend refactor (part 2)

lindsayplatt opened the pull request:
- [x] New workflow in each graphics call 
- [x] Combine logic in set_legend_args - don't need to separate lines/pts as much
- [x] Combine polygon/rect logic 
- [x] Move legend.args into object>>legend (they should not be top-level) 
- [x] ~~Use `lazy_eval` instead of `quote`?~~ _more complicated than quote for this case_
- [x] ~~Add `lazy_eval` stuff from set_window_args to handle expressions~~ _doesn't need anything to handle expressions because it is only taking graphical parameters such as `lwd, lty, col` for the legend and ignores everything else._ 
- [x] Fix: using legend.name in error_bars gives a legend item for each error type (y.high, y.low, x.high, x.low)
- [x] ~~Fix `is.na(data.var)` warning from lines 141-151 of calc_views.R~~ _made into a separate issue since it is not caused by legend() - issue_ #256
- [x] Fix warning from having legend arguments in the `to.gsplot`variable in set_window_args
- [x] Fix/look into warning coming from base graphics
- [x] documentation for which args to use in legend vs graphics call
- [x] tests to check that overall legend arg static list (legend.R) & fun args static list (set_args.R) are what we expect

Weird thing with having error_bars use legend.name: for each part of the error bar (y.high, y.low, etc), an arrows call is made - therefore, a separate legend.name is added. This has two legend items:

``` r
gs <- gsplot()
gsNew <- points(gs, y=c(3,1,2), x=1:3, xlim=c(0,NA),ylim=c(0,NA),
                col="blue", pch=18, legend.name="Points")
gsNew <- lines(gsNew, c(3,4,3), c(2,4,6), legend.name="Lines")
gsNew <- abline(gsNew, b=1, a=0, legend.name="1:1")
gsNew <- legend(gsNew, location = "topleft",title="Awesome!")
gsNew <- grid(gsNew)
gsNew <- error_bar(gsNew, 1:3, y=c(3,1,2), y.high=c(0.5,0.25,1), y.low=0.1, legend.name="errors")
gsNew <- error_bar(gsNew, x=1:3, y=c(3,1,2), x.low=c(.2,NA,.2), x.high=.2, col="red",lwd=3)
gsNew <- title(gsNew, "Graphing Fun")
gsNew 
```

Warnings:

```
In is.na(data.var) :
  is.na() applied to non-(list or vector) of type 'NULL'

In (function (xleft, ybottom, xright, ytop, density = NULL, angle = 45,  :
  "text.font" is not a graphical parameter

In x1 + dx :
  longer object length is not a multiple of shorter object length
```

Commit:
remove default title

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
type was changed somehow (need an lty if type="b" or type="o" for points and need pch if its for lines)

Commit:
remove title default

Commit:
Work in progress for legend

Commit:
merging conflicts

Merge remote-tracking branch 'upstream/master'

Conflicts:
	R/lines.R
	R/points.R

Commit:
more wip for legend

Commit:
polygon/rect added + lots warnings (but not errors!)

Commit:
allow different views in same legend

Commit:
renamed variable - call correct one in final legend()

Commit:
cleanup/rearranging old code that needs to be incorporated

Commit:
added type="b", etc

Commit:
add character linetypes

Commit:
dynamically set columns for above/below legends

Commit:
multiple legends work (last example for legend)

Commit:
clean up/adding more plot elements (set_args)

Commit:
added the rest of the plot features that could be in legend

Commit:
correctly handle views - can distinguish between side=4, side=c(3,4) and side=c(1,4)

Commit:
more cleaning, some set_args corrections for "type" logic

Commit:
new workflow for graphics calls

Commit:
combining logic, put legend.args inside object>>legend

Commit:
error bars can be part of legend

Commit:
legend corrections from code review & set_legend_args changes

Commit:
more set_args cleanup

lindsayplatt commented:
Transparent legend:

``` r
gsNew <- gsplot() %>% 
    rect(0,0,4,2, col="lightgreen", border=NA, legend.name="Shape") %>%
    points(y=c(3,1,2), x=1:3, xlim=c(0,NA),ylim=c(0,NA),
           col="blue", pch=18, legend.name="Points") %>% 
    lines(c(3,4,3), c(2,4,6), legend.name="Lines") %>% 
    abline(b=1, a=0, legend.name="1:1") %>% 
    legend(location = "topright", text.col="white", 
           box.lwd=2, ncol=2, bg=NA) %>% 
    grid() %>% 
    error_bar(1:3, y=c(3,1,2), y.high=c(0.5,0.25,1), y.low=0.1) %>% 
    error_bar(x=1:3, y=c(3,1,2), x.low=c(.2,NA,.2), x.high=.2, col="red",lwd=3) %>% 
    title("Graphing Fun")
gsNew
```

![image](https://cloud.githubusercontent.com/assets/13220910/10205042/a04befba-6784-11e5-8e73-f4f9d4a62bda.png)

lindsayplatt commented:
``` r
gs <- gsplot()  %>% 
    rect(1,5,2,7, col="pink", border=NA, legend.name="rectangle", text.font=3) %>%
    points(1:5, 1:5, pch=5, col="blue", legend.name="points") %>%
    lines(3:7, 4:8, col="darkgreen", legend.name="line") %>%
    error_bar(x=4, y=4, y.high=0.5, y.low=0.7, epsilon=0.5, legend.name="Errors") %>% 
    legend(x="bottomright", bg="grey", box.lty="dashed", box.lwd=3)
gs
```

![image](https://cloud.githubusercontent.com/assets/13220910/10205541/5b077ebc-6787-11e5-9c70-1ee0363e61aa.png)

ldecicco-USGS commented:
<snifff > it's sooo pretty <\sniff >

I assume we only want 1 line for `error_bar`'s? 

Edit...I can't surround my text with humorous tags....github formats it out.

lindsayplatt commented:
yep, I've got that in my list of things left to do!

Commit:
don't use legend.name more than once in error_bars

Commit:
Fix warning from having legend arguments in the to.gsplot variable in set_window_args

Commit:
fix `x1+d1` warning - seg.len is an overall legend arg (not a vector)

Commit:
improved roxygen comments for using legend

Commit:
added test for static lists of overall and function-specific legend arguments

lindsayplatt commented:
Duplicated legend items for error bars is fixed:

``` r
gs <- gsplot()  %>% 
    rect(1,5,2,7, col="pink", border=NA, legend.name="rectangle", text.font=3) %>%
    points(1:5, 1:5, pch=5, col="blue", legend.name="points") %>%
    lines(3:7, 4:8, col="darkgreen", legend.name="line") %>%
    error_bar(x=4, y=4, y.high=0.5, y.low=0.7, epsilon=0.5, legend.name="Errors") %>% 
    legend(x="bottomright", bg="grey", box.lty="dashed", box.lwd=3)
gs
```

![image](https://cloud.githubusercontent.com/assets/13220910/10232060/f470384a-684b-11e5-80e7-208e8d1deefe.png)

lindsayplatt commented:
@ldecicco-USGS @jread-usgs @jiwalker-usgs @mhines-usgs 

Here is my rework of `legend` from the comments on Monday's code review. The new commits start at "new workflow for graphics calls". I have also included a couple of graphs that show off how easy it is to add legends in gsplot.

jiwalker-usgs commented on tests/testthat/test-legend.R:
```diff
@@ -23,6 +23,23 @@ test_that("gsplot legend works", {
   
+})
+
+test_that("gsplot legend args are still the same", {
```
:+1:

Commit:
missing some logic for type

Commit:
remove legend from yaml (not used anyways)

Commit:
more legend tests

Commit:
lty and lwd should be NA for polygons and rectangles in legend, even if specified

Commit:
lty and lwd should be NA for polygons and rectangles in legend, even if specified

Commit:
update for tests

Commit:
merging conflicts

Merge branch 'master' of https://github.com/lindsaycarr/gsplot

Conflicts:
	R/set_args.R

jordansread commented:
still a wip?

lindsayplatt commented:
just waiting for a :+1: from reviewers I tagged :)

jiwalker-usgs commented:
:+1:

mhines-usgs commented:
I haven't pulled it down and played yet, but dang that's pretty! will grab and see what happens! :+1:

jordansread commented:
:+1:

lindsayplatt commented:
This should fix #244 #226 #219 #200 #160 #213
